### PR TITLE
`eslint`: add `parent` property to node

### LINKED ISSUE #35701: [@types/eslint] Type `Node` should have property `parent`
Currently the type `Node` is imported from `@types/estree`. But that type lacks the property `parent`, which is added by eslint. Since `node.parent` is frequently used. This is a high impact bug.

The only solution I find is to vendor the a modified copy of `@types/estree` in `@types/eslint`.
 
@pmdartus @j-f1 @saadq @mysticatea

### PR DESCRIPTION
**Pull request by OliverJAsh**
Fixes https://github.com/DefinitelyTyped/DefinitelyTyped/issues/35701

---

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test YOUR_PACKAGE_NAME`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## eslint ([<kbd>unpkg</kbd>](https://unpkg.com/browse/eslint@latest/))
was missing the following properties:
1. linter
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 5d1741a96f1641a6c087f11f4832b1b61c0d99ea
</p>

File: types/eslint/index.d.ts [hunk 1]
```diff
@@ -246,92 +246,95 @@ export namespace Rule {
...
+        LabeledStatement?: (node: ESTree.LabeledStatement & NodeExtensions) => void;
+        Literal?: (node: ESTree.Literal & NodeExtensions) => void;
+        LogicalExpression?: (node: ESTree.LogicalExpression & NodeExtensions) => void;
+        MemberExpression?: (node: ESTree.MemberExpression & NodeExtensions) => void;
+        MetaProperty?: (node: ESTree.MetaProperty & NodeExtensions) => void;
+        MethodDefinition?: (node: ESTree.MethodDefinition & NodeExtensions) => void;
+        NewExpression?: (node: ESTree.NewExpression & NodeExtensions) => void;
+        ObjectExpression?: (node: ESTree.ObjectExpression & NodeExtensions) => void;
+        ObjectPattern?: (node: ESTree.ObjectPattern & NodeExtensions) => void;
+        Program?: (node: ESTree.Program & NodeExtensions) => void;
```
Comment (bradzacher) [inline types/eslint/index.d.ts:295]: A Program node will never have a parent.

File: types/eslint/index.d.ts [hunk 2]
```diff
@@ -246,92 +246,95 @@ export namespace Rule {
...
+        UpdateExpression?: (node: ESTree.UpdateExpression & NodeExtensions) => void;
+        VariableDeclaration?: (node: ESTree.VariableDeclaration & NodeExtensions) => void;
+        VariableDeclarator?: (node: ESTree.VariableDeclarator & NodeExtensions) => void;
+        WhileStatement?: (node: ESTree.WhileStatement & NodeExtensions) => void;
+        WithStatement?: (node: ESTree.WithStatement & NodeExtensions) => void;
+        YieldExpression?: (node: ESTree.YieldExpression & NodeExtensions) => void;
+    }
+
+    interface NodeExtensions { parent: Node; }
+    type Node = ESTree.Node & NodeExtensions;
```
Comment (bradzacher) [inline types/eslint/index.d.ts:320]: This will erroneously add `parent` to `Program`

**Review by bradzacher — CHANGES_REQUESTED** [review]
Program is the only node that cannot have a parent because its the root node!

File: types/eslint/index.d.ts (hunk 1 above)
Comment (OliverJAsh) [reply types/eslint/index.d.ts]: Thanks, fixed!

**Review by bradzacher — APPROVED** [review]

**Comment by OliverJAsh** [comment]
Ready to merge

**Pull request merged**